File: Leetcode/0098-validate-binary-search-tree/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        """Time O(n), Space O(h)."""
        def dfs(n, lo, hi):
            if not n: return True
            if n.val <= lo or n.val >= hi: return False
            return dfs(n.left, lo, n.val) and dfs(n.right, n.val, hi)
        return dfs(root, float('-inf'), float('inf'))


def build_tree(arr):
    if not arr or arr[0] is None: return None
    root = TreeNode(arr[0])
    q = [root]; i = 1
    while q and i < len(arr):
        n = q.pop(0)
        if i < len(arr) and arr[i] is not None: n.left = TreeNode(arr[i]); q.append(n.left)
        i += 1
        if i < len(arr) and arr[i] is not None: n.right = TreeNode(arr[i]); q.append(n.right)
        i += 1
    return root


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    cases = [
        ("Valid 1", [2, 1, 3], True),
        ("Invalid 1", [5, 1, 4, None, None, 3, 6], False),
        ("Single", [1], True),
        ("Empty", [], True),
        ("Equal not allowed", [1, 1], False),
        ("Deep valid", [4, 2, 6, 1, 3, 5, 7], True),
        ("Deep invalid", [10, 5, 15, None, None, 6, 20], False),
    ]
    for n, t, exp in cases:
        got = sol.isValidBST(build_tree(t))
        if got == exp: print(f"PASS: {n}"); passed += 1
        else: print(f"FAIL: {n} got={got}"); failed += 1
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")

File: Leetcode/0098-validate-binary-search-tree/solution.go
package main

import "fmt"

type TreeNode struct {
	Val         int
	Left, Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	var dfs func(n *TreeNode, lo, hi int64) bool
	dfs = func(n *TreeNode, lo, hi int64) bool {
		if n == nil {
			return true
		}
		v := int64(n.Val)
		if v <= lo || v >= hi {
			return false
		}
		return dfs(n.Left, lo, v) && dfs(n.Right, v, hi)
	}
	return dfs(root, -1<<62, 1<<62)
}

func buildTree(arr []interface{}) *TreeNode {
	if len(arr) == 0 || arr[0] == nil {
		return nil
	}
	root := &TreeNode{Val: arr[0].(int)}
	q := []*TreeNode{root}
	i := 1
	for len(q) > 0 && i < len(arr) {
		n := q[0]
		q = q[1:]
		if i < len(arr) && arr[i] != nil {
			n.Left = &TreeNode{Val: arr[i].(int)}
			q = append(q, n.Left)
		}
		i++
		if i < len(arr) && arr[i] != nil {
			n.Right = &TreeNode{Val: arr[i].(int)}
			q = append(q, n.Right)
		}
		i++
	}
	return root
}

func main() {
	passed, failed := 0, 0
	test := func(name string, got, exp bool) {
		if got == exp {
			fmt.Printf("PASS: %s\n", name)
			passed++
		} else {
			fmt.Printf("FAIL: %s got=%v\n", name, got)
			failed++
		}
	}
	cases := []struct {
		name string
		arr  []interface{}
		want bool
	}{
		{"Valid 1", []interface{}{2, 1, 3}, true},
		{"Invalid 1", []interface{}{5, 1, 4, nil, nil, 3, 6}, false},
		{"Single", []interface{}{1}, true},
		{"Empty", []interface{}{}, true},
		{"Equal not allowed", []interface{}{1, 1}, false},
		{"Deep valid", []interface{}{4, 2, 6, 1, 3, 5, 7}, true},
		{"Deep invalid", []interface{}{10, 5, 15, nil, nil, 6, 20}, false},
	}
	for _, c := range cases {
		test(c.name, isValidBST(buildTree(c.arr)), c.want)
	}
	fmt.Printf("\nResults: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}
